atc/db/build_in_memory_check.go

```diff
@@ -0,0 +1,560 @@
+	}
+	return b.checkable != nil && b.checkable.BuildSummary() != nil && b.checkable.BuildSummary().PublicPlan != nil
+}
+
+func (b *inMemoryCheckBuild) PublicPlan() *json.RawMessage {
+	if b.plan.ID != "" {
+		return b.plan.Public()
+	}
+
+	resource, ok := b.checkable.(Resource)
+	if !ok || resource.BuildSummary() == nil || resource.BuildSummary().PublicPlan == nil {
+		return nil
+	}
+	bytes, err := json.Marshal(resource.BuildSummary().PublicPlan)
+	if err != nil {
+		return nil
+	}
+	m := json.RawMessage(bytes)
+	return &m
+}
+
+// ResourceCacheUser return no-user because a check build may only generate a image
+// resource and image resource will be cached by SaveImageResourceVersion.
+func (b *inMemoryCheckBuild) ResourceCacheUser() ResourceCacheUser {
+	if !b.running {
+		panic("not-implemented")
+	}
+	return NoUser()
+}
+
+func (b *inMemoryCheckBuild) ContainerOwner(planId atc.PlanID) ContainerOwner {
+	if !b.running {
+		panic("not-implemented")
+	}
+
+	if b.id == 0 {
+		panic("in-memory-build-not-running-yet")
+	}
+
+	return NewInMemoryCheckBuildContainerOwner(b.id, planId, b.TeamID())
+}
+
+// SaveImageResourceVersion does nothing. Because if a check use a custom resource
+// type, the resource type image's resource cache id will be set in the resource's
+// resource config as resource_cache_id, so that the image's resource cache will not
+// be GC-ed. As checks run every minute, the resource_config's last_referenced time
+// keeps updated, then the image's resource cache will be always retained.
+func (b *inMemoryCheckBuild) SaveImageResourceVersion(cache UsedResourceCache) error {
+	return nil
+}
+
+func (b *inMemoryCheckBuild) AllAssociatedTeamNames() []string {
+	if b.cacheAssociatedTeams != nil {
+		return b.cacheAssociatedTeams
+	}
+
+	rows, err := psql.Select("distinct(t.name)").
+		From("resources r").
+		LeftJoin("pipelines p on r.pipeline_id = p.id").
+		LeftJoin("teams t on p.team_id = t.id").
+		Where(sq.Eq{"r.resource_config_scope_id": b.checkable.ResourceConfigScopeID()}).
+		RunWith(b.conn).
+		Query()
+	if err != nil {
+		return []string{b.checkable.TeamName()}
+	}
+	defer Close(rows)
+
+	var teamNames []string
+	for rows.Next() {
+		var teamName string
+		err := rows.Scan(&teamName)
+		if err != nil {
+			return teamNames
+		}
+		teamNames = append(teamNames, teamName)
+	}
+	b.cacheAssociatedTeams = teamNames
+
+	return b.cacheAssociatedTeams
+}
+
+func (b *inMemoryCheckBuild) initDbStuff(tx Tx) error {
+	var nextBuildId int
+	err := psql.Select("nextval('builds_id_seq'::regclass)").RunWith(tx).QueryRow().Scan(&nextBuildId)
+	if err != nil {
+		return err
+	}
+	b.id = nextBuildId
+
+	for _, cachedEv := range b.cacheEvents {
+		err := b.saveEvent(tx, cachedEv)
+		if err != nil {
+			return err
+		}
+	}
+
+	b.dbInited = true
+	b.cacheEvents = []atc.Event{}
+
+	return nil
+}
+
+// === No implemented functions ===
+
+func (b *inMemoryCheckBuild) CreateTime() time.Time                        { panic("not-implemented") }
```

Comment:
Good point. Let me try to figure it out.

Reply:
@clarafu  First, I fully buy in your idea of smaller interface. And I just did some quick experiment, and realized that would be more harder than I expected. (But a good thing is, during the experiment, I found that I missed implementing some functions for in-memory build.)

Basically, `db.Build` has two branches of usages:
* Used in `engine` to run a build
* Used in `api` to view or manage a build

Some functions are used only in `engine`, and some are only used on `api`. But in-memory build are called from both branches, it's only logically that some functions of `in-memory-build` should never be called because `in-memory-build` are only Lidar-triggered check builds, thus ends up that `in-memory-build` still need to implement full set of `Build` functions.

So changing to use smaller interfaces would be too hard to implement, especially, given this PR has been too big, I don't feel comfortable to make the change.

To address your concern, I thinking that, maybe we can revise panic messages. Instead of "not-implemented", "in-memory-build-cannot-set-comment" might be better.

If you really don't like `panic`, I can change to return error like `return nil, errors.New("in memory build cannot set comment")` if a function returns an error.

Reply:
Ya definitely making this change would involve a lot more work, so I think it's fine to leave it maybe for a separate refactor after this pr is merged in.

I think I would feel more comfortable with returning errors and having the components deal with the errors rather than panic and crashing the application. What do you think?

Reply:
Yeah, if you look my latest commit, I have deleted all `panic`.